### ISSUE TITLE
Assert less-than-pointer allocation to be leaf

### ISSUE THREAD

**Issue by leirocks** [Microsoft/ChakraCore#880]
I hit some less-than-pointer (like 3 bytes on x86) allocation with leaf bit off in both chakra and memGC
we should fix the caller(runtime code) to make such allocation is leaf, the benefits are:
1. reduce addresses to scan
2. reduce false positive in marking

we can add an assert in recycler to find out such allocations

**Comment by agarwal-sandeep** [comment]
@leirocks is this still an issue?

//cc @kfarnung @dilijev

**Comment by leirocks** [comment]
yes, for less-than-pointer size allocation should be allocated with leaf, because it can't contain pointer. or even eliminate those allocation if possible (like merge those allocation to some bigger struct which have same life time)

we can add assertion to rule out those allocations. I know there are few in Promise, and couple other places
